README.md
```
--- 1 ---
x₁ x₂ t₁ t₂ Y Z
0.15 0.40 1.40 2.60 0.25±0.00 2.40±0.20
0.15 0.50 1.50 3.10 0.35±0.00 3.68±0.23
0.15 0.70 1.40 3.70 0.55±0.00 5.87±0.26
0.15 0.90 1.30 4.00 0.75±0.00 7.16±0.28
0.15 1.10 1.40 4.60 0.95±0.00 9.60±0.32
σa = 0.00195
a = 0.099±0.004
ε = 3.916%
--- 2 ---
sinα t₁±Δt₁ t₂±Δt₂ ⟨a⟩±Δa
0.0128 1.38±0.08 4.58±0.08 0.10±0.00
0.0244 1.00±0.07 3.40±0.07 0.18±0.01
0.0346 0.80±0.07 2.70±0.07 0.29±0.02
0.0449 0.70±0.07 2.30±0.09 0.40±0.04
0.0564 0.60±0.07 2.00±0.07 0.52±0.04
A = -0.0440
B = 9.8409
σg = 0.4777
Δg = 0.9554
g = 9.84±0.96
ε = 9.71%
|g_real - g| = 0.0214
```


CHANGELOG.md
# Revision history for physics3-lab2-app

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


app/Task1.hs
{-# LANGUAGE OverloadedStrings #-}
module Task1 (main) where

import           Control.Monad (forM_)
import           Fmt
import           GHC.Float     (int2Double)
import qualified Numeric.AD    as D

h₀ = 0.191
h₀' = 0.193

yF x₁ x₂ = x₂ - x₁
zF t₁ t₂ = (t₂^2 - t₁^2)/2

δY :: Double -> Double -> Double
δY x₁ x₂ = sqrt $ (pd₁ * δx₁)^2 + (pd₂ * δx₂)^2
  where
    δx₁ = 2 / 3 * 5 / 1000
    δx₂ = 2 / 3 * 5 / 1000

    pd₁ = D.diff (\x1 -> yF x1 (D.auto x₂)) x₁
    pd₂ = D.diff (\x2 -> yF (D.auto x₁) x2) x₂

δZ :: Double -> Double -> Double
δZ t₁ t₂ = sqrt $ (pd₁ * δt₁)^2 + (pd₂ * δt₂)^2
  where
    δt₁ = 2 / 3 * 0.1
    δt₂ = 2 / 3 * 0.1

    pd₁ = D.diff (\t1 -> zF t1 (D.auto t₂)) t₁
    pd₂ = D.diff (\t2 -> zF (D.auto t₁) t2) t₂

dataset =
  [ ((0.15, 0.4), (1.4, 2.6))
  , ((0.15, 0.5), (1.5, 3.1))
  , ((0.15, 0.7), (1.4, 3.7))
  , ((0.15, 0.9), (1.3, 4.0))
  , ((0.15, 1.1), (1.4, 4.6))
  ]

ys = [ (yF x₁ x₂, δY x₁ x₂) | ((x₁, x₂), _) <- dataset]
zs = [ (zF t₁ t₂, δZ t₁ t₂) | (_, (t₁, t₂)) <- dataset]

pprintNum :: Int -> Double -> Double -> String
pprintNum prec a δa = "" +| (fixedF prec a) |+ "±" +| (fixedF prec δa) |+ ""

theA = sum (zipWith (*) (map fst ys) (map fst zs)) / sum (map (^2) (map fst zs))

σa = sqrt $ s₁ / (cnt - 1) / s₂
  where
    s₁ = sum $ zipWith (\y z -> (y - theA * z)^2) (map fst ys) (map fst zs)
    s₂ = sum $ map (^2) (map fst zs)
    cnt = int2Double $ length ys

δa = 2 * σa

main :: IO ()
main = do
  putStrLn "x₁ x₂ t₁ t₂ Y Z"
  let prec = 2
  forM_ (zip3 dataset ys zs) $ \(((x₁, x₂), (t₁, t₂)), (y, δy), (z, δz)) -> do
    putStrLn $ "" +| (fixedF prec x₁) |+ " " +| (fixedF prec x₂) |+
              " " +| (fixedF prec t₁) |+ " " +| (fixedF prec t₂) |+
              " " +| pprintNum prec y δy |+ " " +| pprintNum prec z δz |+ ""
  putStrLn $ "σa = " +| (fixedF 5 σa) |+ ""
  putStrLn $ "a = " +| pprintNum 3 theA δa |+ ""
  putStrLn $ "ε = " +| (fixedF 3 (δa / theA * 100)) |+ "%"


app/Task2.hs
{-# LANGUAGE OverloadedRecordDot #-}
{-# LANGUAGE OverloadedStrings   #-}
module Task2 (main) where

import           Control.Monad (forM_)
import           Data.List     (zip4)
import           Fmt
import           GHC.Float     (int2Double)
import qualified Numeric.AD    as D

data Measurement
  = Measurement
  { h  :: !Double
  , h' :: !Double
  , ts :: ![(Double, Double)]
  }

ms :: [Measurement]
ms =
  [ Measurement (200 / 1000) (192 / 1000)
    [ (1.4, 4.6)
    , (1.4, 4.6)
    , (1.3, 4.5)
    , (1.4, 4.6)
    , (1.4, 4.6)
    ]
  , Measurement (210 / 1000) (193 / 1000)
    [ (1.0, 3.4)
    , (1.0, 3.4)
    , (1.0, 3.4)
    , (1.0, 3.4)
    , (1.0, 3.4)
    ]
  , Measurement (218 / 1000) (193 / 1000)
    [ (0.8, 2.7)
    , (0.8, 2.7)
    , (0.8, 2.7)
    , (0.8, 2.7)
    , (0.8, 2.7)
    ]
  , Measurement (227 / 1000) (194 / 1000)
    [ (0.7, 2.3)
    , (0.7, 2.4)
    , (0.7, 2.3)
    , (0.7, 2.2)
    , (0.7, 2.3)
    ]
  , Measurement (237 / 1000) (195 / 1000)
    [ (0.6, 2.0)
    , (0.6, 2.0)
    , (0.6, 2.0)
    , (0.6, 2.0)
    , (0.6, 2.0)
    ]
  ]

h₀ = 191 / 1000
h₀' = 193 / 1000

x = 0.22
x' = 1

x₁ = 0.15
x₂ = 1.1

sinα :: Measurement -> Double
sinα m = ((m.h - h₀) - (m.h' - h₀')) / (x' - x)

sinαs = map sinα ms

mean :: Fractional a => [a] -> a
mean xs = s / l
  where
    (s, l) = foldr (\x (s', l') -> (s' + x, l' + 1)) (0, 0) xs

σ :: [Double] -> Double
σ xs = sqrt $ s / l / (l - 1)
  where
    m = mean xs
    l = int2Double $ length xs
    s = sum $ [ (x - m)^2 | x <- xs ]

tαN = 2.13
δT = 2 / 3 * 0.1

processT :: [Double] -> (Double, Double)
processT ts = (m, sqrt (δt^2 + δT^2))
  where
    m = mean ts
    σt = σ ts
    δt = tαN * σt

as_ = do
  (Measurement _ _ ts) <- ms
  let mt₁ = mean $ map fst ts
  let mt₂ = mean $ map snd ts
  pure $ 2 * (x₂ - x₁) / (mt₂^2 - mt₁^2)

δas = do
  ((Measurement _ _ ts), ma) <- zip ms as_
  let δx = 5 / 1000
  let (mt₁, δt₁) = processT $ map fst ts
  let (mt₂, δt₂) = processT $ map snd ts
  pure $ ma * sqrt (
    (δx^2 + δx^2)/(x₂ - x₁)^2
    + 4 * ((mt₁ * δt₁)^2 + (mt₂ * δt₂)^2) / (mt₂^2 - mt₁^2)^2)


theB = (s₁ - s₂ * s₃ / cnt) / (s₄ - s₃^2 / cnt)
  where
    s₁ = sum $ zipWith (*) as_ sinαs
    s₂ = sum as_
    s₃ = sum sinαs
    s₄ = sum $ map (^2) sinαs
    cnt = int2Double $ length ms

theA = (sum as_ - theB * (sum sinαs)) / (int2Double (length ms))

theD = sum (map (^2) sinαs) - (sum sinαs)^2 / (int2Double $ length ms)

σg = sqrt $ s₁ / (theD * (cnt - 2))
  where
    cnt = int2Double $ length ms
    s₁ = sum $ zipWith (\a s -> (a - (theA + theB * s))^2) as_ sinαs

δg = 2 * σg

g_real = 9.8195

pprintNum :: Int -> Double -> Double -> String
pprintNum prec a δa = "" +| (fixedF prec a) |+ "±" +| (fixedF prec δa) |+ ""

main :: IO ()
main = do
  let prec = 2
  putStrLn "sinα t₁±Δt₁ t₂±Δt₂ ⟨a⟩±Δa"
  forM_ (zip4 ms sinαs as_ δas) $ \((Measurement _ _ ts), s, a, δa) -> do
    let (mt₁, δt₁) = processT $ map fst ts
    let (mt₂, δt₂) = processT $ map snd ts
    putStrLn $ "" +| (fixedF 4 s) |+ " " +| pprintNum prec mt₁ δt₁ |+
              " " +| pprintNum prec mt₂ δt₂ |+
              " " +| pprintNum prec a δa |+ ""
  putStrLn $ "A = " +| fixedF 4 theA |+ ""
  putStrLn $ "B = " +| fixedF 4 theB |+ ""
  putStrLn $ "σg = " +| fixedF 4 σg |+ ""
  putStrLn $ "Δg = " +| fixedF 4 δg |+ ""
  putStrLn $ "g = " +| pprintNum prec theB δg |+ ""
  putStrLn $ "ε = " +| fixedF prec (δg / theB * 100) |+ "%"
  putStrLn $ "|g_real - g| = " +| fixedF 4 (abs (g_real - theB)) |+ ""
